lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,282 @@
+  }
+
+  async #watchFolder(folder) {
+    const { opendir } = lazyLoadFsPromises();
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        if (this.#closed) {
+          break;
+        }
+
+        const f = path.join(folder, file.name);
+
+        if (!this.#files.has(f)) {
+          this.#files.set(f, file);
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isDirectory()) {
+            await this.#watchFolder(f);
+          } else {
+            this.#watchFile(f);
+          }
+        }
+      }
+    } catch (error) {
+      this.emit('error', error);
+    }
+  }
+
+  #watchFile(file) {
+    if (this.#closed) {
+      return;
+    }
+
+    const { watchFile } = lazyLoadFsSync();
+    const existingStat = this.#files.get(file);
+
+    watchFile(file, {
+      persistent: this.#options.persistent,
+    }, (statWatcher, previousStatWatcher) => {
+      if (existingStat && !existingStat.isDirectory() &&
+        statWatcher.nlink !== 0 && existingStat.mtime?.getTime() === statWatcher.mtime?.getTime()) {
+        return;
+      }
+
+      this.#files.set(file, statWatcher);
+
+      if (statWatcher.birthtimeMs === 0 && previousStatWatcher.birthtimeMs !== 0) {
+        // The file is now deleted
+        this.#files.delete(file);
+        this.emit('change', 'rename', this.#getPath(file));
+
+        if (statWatcher.isDirectory()) {
+          this.#unwatchFolder(file);
+        }
+      } else {
+        this.emit('change', 'change', this.#getPath(file));
+
+        if (statWatcher.isDirectory()) {
+          this.#watchFolder(file);
+        }
+      }
+    });
+  }
+
+  [kFSWatchStart](filename) {
+    filename = path.resolve(getValidatedPath(filename));
+
+    try {
+      const file = lazyLoadFsSync().statSync(filename);
+
+      this.#rootPath = filename;
+      this.#closed = false;
+      this.#files.set(filename, file);
+      this.#watchingFile = file.isFile();
+
+      if (file.isDirectory()) {
+        PromisePrototypeThen(
+          traverse(filename, this.#files),
+          () => {
+            for (const f of this.#files.keys()) {
+              this.#watchFile(f);
+            }
+          },
+        );
+      } else {
+        this.#watchFile(filename);
+      }
+    } catch (error) {
+      if (error.code === 'ENOENT') {
+        error.filename = filename;
+        throw error;
+      }
+    }
+
+  }
+
+  ref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('ref');
+  }
+
+  unref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('unref');
+  }
+
+  [SymbolAsyncIterator]() {
+    const self = this;
+    const stream = new Readable({
+      destroy(error, callback) {
+        self.close();
+        if (error.name !== 'AbortError') {
+          callback(error);
+        } else {
+          callback(null);
+        }
```

Comment:
I don't think we should swallow the error here. The only reason we would have an `AbortError` here would be if `this.#options.signal` is an aborted `AbortSignal`, and we would want to throw that error to the user in this case.
```suggestion
        callback(error);
```

Reply:
Well, this change throws an error if `break` is called inside a for await loop, since it terminates the abort signal of the async iterator.

Reply:
We need to find a way to change that somehow, without swallowing the error. At the very least it would need to have a TODO asking to fix it later, but I think it should be possible to fix it in this PR, we just need to find what's triggering the abort of the `AbortSignal` and disable it somehow.

Reply:
I'm going to add a TODO for now.